### PR TITLE
getStatus command from grid should not be queued

### PR DESCRIPTION
**Pull request by yaphatak**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Appium?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla.js.foundation/appium/appium)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: lib/appium.js [hunk 1]
```diff
@@ -283,7 +287,7 @@ class AppiumDriver extends BaseDriver {
 // help decide which commands should be proxied to sub-drivers and which
 // should be handled by this, our umbrella driver
 function isAppiumDriverCommand (cmd) {
-  return !isSessionCommand(cmd) || cmd === "deleteSession";
+  return (!isSessionCommand(cmd) || cmd === "deleteSession") && cmd !== 'getStatus';
```
Comment (mykola-mokhnach) [inline lib/appium.js]: I'd rather change it to `!isSessionCommand(cmd) || ['deleteSession', 'getStatus'].indexOf(cmd) >= 0`

File: lib/appium.js [hunk 2]
```diff
@@ -257,7 +257,11 @@ class AppiumDriver extends BaseDriver {
     if (isAppiumDriverCommand(cmd)) {
       return super.executeCommand(cmd, ...args);
     }
-
+    // getStatus command should not be put into queue. If we do it as part of executionCommand, it will be added to queue.
+    // There will be lot of status commands in queue during createSession command, as createSession can take up to or more than a minute.
+    if (cmd === 'getStatus') {
```
Comment (mykola-mokhnach) [inline lib/appium.js]: is condition ever reached since isAppiumDriverCommand always returns _true_ for `cmd === 'getStatus'`?

Comment (imurchie) [inline lib/appium.js]: I would rather just handle `getStatus` at the start of this function, and not worry about it in `isAppiumDriverCommand`.

**Review by imurchie — COMMENTED** [review]
How do we handle individual drivers that implement their own `getStatus` functions, in this scenario?

**Comment by yaphatak** [comment]
done

File: lib/appium.js [hunk 3]
```diff
@@ -254,10 +254,14 @@ class AppiumDriver extends BaseDriver {
...
   async executeCommand (cmd, ...args) {
+    // getStatus command should not be put into queue. If we do it as part of super.executionCommand, it will be added to queue.
+    // There will be lot of status commands in queue during createSession command, as createSession can take up to or more than a minute.
+    if (cmd === 'getStatus') {
+      return await this.getStatus();
+    }
     if (isAppiumDriverCommand(cmd)) {
       return super.executeCommand(cmd, ...args);
     }
-
```
Comment (mykola-mokhnach) [inline lib/appium.js]: can we keep that line break?

**Review by mykola-mokhnach — APPROVED** [review]

File: lib/appium.js [hunk 4]
```diff
@@ -254,6 +254,11 @@ class AppiumDriver extends BaseDriver {
   }
 
   async executeCommand (cmd, ...args) {
+    // getStatus command should not be put into queue. If we do it as part of super.executionCommand, it will be added to queue.
```
Comment (jlipps) [inline lib/appium.js]: `super.executeCommand`

**Review by jlipps — COMMENTED** [review]
1 comment on the comment then LGTM

**Comment by yaphatak** [comment]
> How do we handle individual drivers that implement their own getStatus functions, in this scenario?
I think today before this change,  it is not calling individual drivers. I think we have to understand main purpose of `getStatus`, it is to make sure that node is alive. So this functionality should be in Appium driver as it manages `sub-drivers`. 
Its difficult to do that at least with current architecture as you cant call individual driver unless you have instance of that driver and we have instance only when there is instance. Theoretically , appium node can manage multiple drivers (ios, android, windows), in such case its status of appium node that matters and not 'sub-drivers'.

File: lib/appium.js (hunk 4 above)
Comment (yaphatak) [inline lib/appium.js]: done

**Review by jlipps — APPROVED** [review]

**Comment by imurchie** [comment]
That makes sense. Currently the `appium-xcuitest-driver` implements its own command (and logic to bypass the queue).

**Comment by mykola-mokhnach** [comment]
@imurchie do we need to update anything in the xctest driver because of this change?

**Comment by imurchie** [comment]
@mykola-mokhnach No. We could tear a bunch of stuff related to `getStatus` out, but it is not necessary. With this in place the `getStatus` command will _never_ get to the individual drivers. So unless they themselves are being used in a grid or something, there is no reason for them to implement the method.

**Pull request merged**